### PR TITLE
Clamp COCO polygon vertices to bbox bounds (fix #2847) 

### LINKED ISSUE #2847: [BUG] COCO export: Instance Segmentations OOB wrt their bounding box
### Instructions

Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/voxel51/fiftyone/blob/develop/ISSUE_POLICY.md) for information on what types of issues we address.

1. Please fill in this template to ensure a timely and thorough response
2. Place an "x" between the brackets next to an option if it applies. For example:
    -   [x] Selected option
3. **Please delete everything above this line before submitting the issue**

### System information

-   **OS Platform and Distribution** (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
-   **Python version** (`python --version`): 3.10.10
-   **FiftyOne version** (`fiftyone --version`): 0.20.0
-   **FiftyOne installed from** (pip or source): pip

### Commands to reproduce

```
def check_ann(ann):
    bb = ann['bbox']
    segall = ann['segmentation']

    bbxyxy = bb.copy()
    bbxyxy[2] += bbxyxy[0]
    bbxyxy[3] += bbxyxy[1]

    for seg in segall:
        seg = np.array(seg,dtype=np.float64).reshape((len(seg)//2),2)

        #print(seg1)
        orx0 = seg[:,0]<bbxyxy[0]
        orx1 = seg[:,0]>bbxyxy[2]
        ory0 = seg[:,1]<bbxyxy[1]
        ory1 = seg[:,1]>bbxyxy[3]
        if np.any(orx0):
            print(f'x0, {seg[orx0,0]} against {bbxyxy[0]}')
        if np.any(orx1):
            print(f'x1, {seg[orx1,0]} against {bbxyxy[2]}')
        if np.any(ory0):
            print(f'y0, {seg[ory0,1]} against {bbxyxy[1]}')
        if np.any(ory1):
            print(f'y1, {seg[ory1,1]} against {bbxyxy[3]}')
    
    return seg, bbxyxy

PATH = '/home/allen/tmp/export_oiv7'

ds = foz.load_zoo_dataset('open-images-v7',max_samples=15,
                          split='validation',label_types=['segmentations'])

ds.export(export_dir=PATH,dataset_type=fo.types.COCODetectionDataset)

ds2 = fo.Dataset.from_dir(dataset_dir=PATH,dataset_type=fo.types.COCODetectionDataset)

JSON = os.path.join(PATH,'labels.json')
with open(JSON) as fh:
    j0 = json.load(fh)

for aidx,ann in enumerate(j0['annotations']):
    print(aidx)
    check_ann(ann)
```

Output:
```
0
x0, [752.5] against 753.651712
1
2
3
y0, [154.5 153.5] against 154.77248
4
5
y0, [350.5 351. ] against 351.99974399999996
6
x0, [266.5 266.5] against 267.327488
y0, [166.5 167.5] against 167.727725
7
y0, [86.5] against 87.560681
8
y0, [71.5 71.5] against 72.42499
9
y0, [468.5] against 469.25214600000004
10
x0, [102.5 103.5] against 103.999488
y0, [86.5 86.5] against 87.999744

... snip ...
```

### Describe the problem

Exporting instance segmentations via COCODetectionDataset can result in segmentations that go out of bounds with respect to their bounding box by up to ~1.5px. This may cause an issue for some training pipelines and/or represent a small loss in accuracy.

I investigated one of the original COCO datasets downloaded from cocodataset.org and this issue was not present. So round-tripping through FO (importing and then re-exporting) likely causes the slight discrepancies.

### What areas of FiftyOne does this bug affect?

-   [ ] `App`: FiftyOne application issue
-   [x] `Core`: Core Python library issue
-   [ ] `Server`: FiftyOne server issue

### Willingness to contribute

The FiftyOne Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the FiftyOne codebase?

-   [ ] Yes. I can contribute a fix for this bug independently
-   [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the FiftyOne community
-   [ ] No. I cannot contribute a bug fix at this time

### PR DESCRIPTION
**Pull request by soham02**
## What changes are proposed in this pull request?

This PR fixes an issue where COCO segmentation polygon vertices could drift slightly outside their corresponding bounding boxes (typically by ~1px) when exporting instance masks.

The root cause was a mismatch between float bounding box coordinates (e.g. 753.65) and pixel-grid–derived polygon vertices. This resulted in invalid COCO annotations where segmentation points exceeded bbox bounds.

The fix clamps polygon vertices to the computed bounding box limits during mask → polygon conversion, preserving the mask shape while ensuring COCO validity.

A regression unit test has been added to verify that exported COCO polygons always lie within their bounding boxes across multiple edge cases (float bbox edges, near-origin bboxes, and tiny boxes).

Fixes #2847.

## How is this patch tested? If it is not, please explain why.

Added a new unit test: test_coco_polygon_clamp.py that constructs synthetic detections with masks and float bounding boxes and asserts that all exported COCO polygon vertices remain within bbox bounds.

Verified locally that the test fails prior to the fix and passes after applying the clamp logic.

Manually validated on a small dataset that COCO exports no longer contain out-of-bounds polygon vertices.

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [X] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

Release note:

Fixes an issue where COCO segmentation polygons exported from instance masks could contain vertices slightly outside their bounding boxes, improving COCO annotation validity for downstream tools.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [X] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

* **Bug Fixes**
  * Segmentation generation during COCO export now respects bounding-box bounds, clamps polygon vertices to bbox and image limits, and ensures non-negative coordinates for valid COCO segmentations; bounding-box rounding is honored.
  * Polyline-derived polygons preserve float precision while staying within image bounds.

* **Tests**
  * Added unit tests covering polygon clamping, bbox-bound handling, rounding edge cases, and coordinate precision.

### CODE REVIEW

File: fiftyone/utils/coco.py [hunk 1]
```diff
@@ -1304,9 +1304,23 @@ def from_label(
...
 
+            bbox_bounds = None
+            if bbox is not None and num_decimals is not None:
+                bbox = [round(p, num_decimals) for p in bbox]
+                bbox_bounds = (
+                    float(bbox[0]),
+                    float(bbox[1]),
+                    float(bbox[0] + bbox[2]),
+                    float(bbox[1] + bbox[3]),
+                )
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply fiftyone/utils/coco.py:1315]: I agree with Rabbit. We assign a value to bbox, so it can’t be None. We can remove this from the conditional.

File: fiftyone/utils/coco.py [hunk 2]
```diff
@@ -2260,32 +2277,59 @@ def _polyline_to_coco_segmentation(polyline, frame_size, iscrowd="iscrowd"):
     for points in polyline.points:
         polygon = []
         for x, y in points:
-            polygon.append(int(x * width))
-            polygon.append(int(y * height))
+            px = x * width
```
Comment (AdonaiVera) [inline fiftyone/utils/coco.py:2280]: I know I suggested this internally @soham02 , but I think it’s better to leave this out of this PR.

The `_polyline_to_coco_segmentation` change switches the output from int to float, which isn’t related to issue #2847 (polylines don’t go through that code path). I’d suggest reverting the function and removing the related test. If we want clamping for polylines, we can tackle that in a separate PR while keeping the `int()` conversion.

File: tests/unittests/utils/test_coco_polygon_clamp.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    assert len(seg) > 0
+    for poly in seg:
+        xs = poly[0::2]
+        ys = poly[1::2]
+        assert all(xmin <= x <= xmax for x in xs)
+        assert all(ymin <= y <= ymax for y in ys)
+
+
+def test_polyline_to_coco_segmentation_preserves_float_and_clamps_to_image():
```
Comment (AdonaiVera) [inline tests/unittests/utils/test_coco_polygon_clamp.py:72]: We can remove this unit test based on the suggestion above.

File: fiftyone/utils/coco.py [hunk 4]
```diff
@@ -2260,32 +2277,59 @@ def _polyline_to_coco_segmentation(polyline, frame_size, iscrowd="iscrowd"):
...
         mask = np.zeros((height, width), dtype=bool)
 
     if detection.get_attribute_value(iscrowd, None):
         return _mask_to_rle(mask)
 
-    return _mask_to_polygons(mask, tolerance)
+    # Compute bbox bounds in absolute pixels so we can clamp polygon vertices.
+    # Why: COCO bboxes can be floats (e.g. 753.65) but contours come from a pixel grid,
+    # which can create ~1px drift outside the bbox. Clamping preserves mask shape while
+    # ensuring COCO validity (segmentation must be within bbox).
```
Comment (AdonaiVera) [inline fiftyone/utils/coco.py:2318]: This is a bit verbose here, we could simplify it to something like:
`# Clamp polygon vertices to bbox bounds to ensure COCO validity`

File: fiftyone/utils/coco.py [hunk 5]
```diff
@@ -2260,32 +2277,59 @@ def _polyline_to_coco_segmentation(polyline, frame_size, iscrowd="iscrowd"):
...
+    if bbox_bounds is None:
+        x, y, w, h = detection.bounding_box  # relative [0..1]
+        bbox_xmin = x * width
```
Comment (AdonaiVera) [inline fiftyone/utils/coco.py:2321]: The intermediate variables (bbox_xmin, etc.) could be inlined to match how from_label computes the same thing, just following more FO code styles.
```
if bbox_bounds is None:
        x, y, w, h = detection.bounding_box
        bbox_bounds = (
            float(x * width),
            float(y * height),
            float(x * width + w * width),
            float(y * height + h * height),
        )
```

File: fiftyone/utils/coco.py [hunk 6]
```diff
@@ -2337,7 +2381,10 @@ def _mask_to_rle(mask):
     return {"counts": counts, "size": list(mask.shape)}
 
 
-def _mask_to_polygons(mask, tolerance):
+# bbox_bounds is optional. When provided, we clamp polygon vertices so they
```
Comment (AdonaiVera) [inline fiftyone/utils/coco.py:2384]: Can you remove this comments, please

File: fiftyone/utils/coco.py [hunk 7]
```diff
@@ -2355,10 +2402,17 @@ def _mask_to_polygons(mask, tolerance):
...
+            xmin, ymin, xmax, ymax = bbox_bounds
+            contour[:, 0] = np.clip(contour[:, 0], xmin, xmax)
+            contour[:, 1] = np.clip(contour[:, 1], ymin, ymax)
+
         segmentation = contour.ravel().tolist()
 
-        # After padding and subtracting 1 there may be -0.5 points
-        segmentation = [0 if i < 0 else i for i in segmentation]
+        # After padding and subtracting 1 there may be negative points
+        segmentation = [max(0, i) for i in segmentation]
```
[the remark this answers is not in the document]
Comment (AdonaiVera) [reply fiftyone/utils/coco.py:2415]: Agree with Rabbit, can you apply this one, please

File: tests/unittests/utils/test_coco_polygon_clamp.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+import unittest.mock as mock
```
Comment (AdonaiVera) [inline tests/unittests/utils/test_coco_polygon_clamp.py:1]: Can you add the docstring here 
```
"""
Tests for fiftyone/utils/coco.py polygon clamping.

| Copyright 2017-2026, Voxel51, Inc.
| `voxel51.com <https://voxel51.com/>`_
|
"""
```

File: fiftyone/utils/coco.py [hunk 9]
```diff
@@ -1329,7 +1343,10 @@ def from_label(
         confidence = label.confidence
 
         if bbox is not None:
-            if num_decimals is not None:
+            # NOTE: bbox rounding for Detection is handled above (before segmentation)
```
Comment (AdonaiVera) [inline fiftyone/utils/coco.py:1346]: Remove this note comment

**Review by brimoor — COMMENTED** [review]
Just flagging: we should reconcile this PR with https://github.com/voxel51/eta/pull/693 before merging, as they are likely related.
